test/suite/hook_demo.py

```diff
@@ -0,0 +1,130 @@
+#
+#   With N == 0, the session.create call merely reports its arguments;
+#   with N == 1, it does an additional session.drop call (which should cause tests to fail);
+#   with N == 2, it does an additional session.create after the drop call (which should work).
+#
+#   Note that notify hooks don't have to simply report, they can call other methods,
+#   set attributes on objects, etc.  For example, one can save the open_cursor
+#   config string as an attribute on the cursor object, and examine it in another
+#   hooked method.
+from __future__ import print_function
+
+import os, sys, wthooks
+from wttest import WiredTigerTestCase
+
+# Print to /dev/tty for debugging, since anything extraneous to stdout/stderr will
+# cause a test error.
+def tty(s):
+    WiredTigerTestCase.tty(s)
+
+# These are the hook functions that are run when particular APIs are called.
+
+# Called to manipulate args for wiredtiger_open
+def wiredtiger_open_args(ignored_self, args):
+    tty('>>> wiredtiger_open, adding cache_size')
+    args = list(args)    # convert from a readonly tuple to a writeable list
+    args[-1] += ',,,cache_size=500M,,,'   # modify the last arg
+    return args
+
+# Called to notify after successful wiredtiger_open
+def wiredtiger_open_notify(ignored_self, ret, *args):
+    tty('>>> wiredtiger_open({}) returned {}'.format(args, ret))
+
+# Called to notify after successful Session.open_cursor
+def session_open_cursor_notify(self, ret, *args):
+    tty('>>> session.open_cursor({}) returned {}, session is {}'.format(args, ret, self))
+
+# Called to replace Session.create
+# We do different things (described above) as indicated by our command line argument.
+def session_create_replace(arg, orig_session_create, session_self, uri, config):
+    tty('>>> session.create({},{}), session is {}'.format(uri, config, session_self))
+    if arg == 0:
+        # Just do a regular create
+        return orig_session_create(session_self, uri, config)
+    elif arg == 1:
+        # Do a regular create, followed by a drop.  This will cause test failures.
+        ret = orig_session_create(session_self, uri, config)
+        # We didn't replace drop, so we can call it as a method
+        tty('>>> session.drop({})'.format(uri))
+        session_self.drop(uri)
+        return ret
+    elif arg == 2:
+        # Do a regular create, followed by a drop, then another create.  Should work.
+        ret = orig_session_create(session_self, uri, config)
+        # We didn't replace drop, so we can call it as a method
+        tty('>>> session.drop({})'.format(uri))
+        session_self.drop(uri)
+        tty('>>> session.create({},{})'.format(uri, config))
+        orig_session_create(session_self, uri, config)
+        return ret
+
+# Every hook file must have one or more classes descended from WiredTigerHook
+# This is where the hook functions are 'hooked' to API methods.
+class DemoHook(wthooks.WiredTigerHook):
+    def __init__(self, arg=0):
```

Comment:
What happens if an hook object, decides to have two functions that use arg? Is the consensus that 2nd function that uses args needs to start from an offset? Might cause a lot of confusion down a complex hook object

Reply:
It's totally legal to have two functions that use arg.  It's up to `DemoHook` coordinate things if it wants something complex.  For example, `DemoHook` could accept an argument like` --hook demo="debug,no_open_cursor_hook,session.create.arg=27,wiredtiger_open.arg=17"`.  Which might mean:
- pass an extra debug arg to all functions
- don't set up the `open_cursor` hook.
- `session.create` gets an extra argument of 27
etc.

You get the idea, it can be arbitrarily complex, and the DemoHook needs to know what to do.  Though I expect if we get to doing stuff like this, we'll make helper functions and evolve as needed.  Let's see where it goes....